### DISCUSSION TITLE
Proposal: "IsRunningOnAzurePipelines" behavior change

### DISCUSSION THREAD

**Discussion by mandalorianbob** [cake-build/cake#3589] [Ideas]
Build System has IsRunningOnAzurePipelines and IsRunningOnAzurePipelinesHosted. 

I would like to propose a change to this. Based on what I see from some of the tests, it appears that this used to differentiate between Cloud and On Prem Azure Dev Ops instances.

However, now, IsRunningOnAzurePipelines returns false even if you're running on Azure Pipelines Cloud, based on Environment.Agent.IsHosted.

This results in _very_ counter intuitive interactions to me, because I'm not on an On Prem AzDO instance, and I never will be. But this is returning different values depending on which build pool I'm running. (Some are self-hosted, some aren't.)

IsRunningOnAzurePipelines should return true regardless of the hosted status of my agent, because it's totally irrelevant to the context that the build agent sets up.

In fact, I'm not sure why this API was set up this way, because the Agent should be fundamentally the same, hosted or not. Unless it's somehow still useful for on prem stuff, I think IsRunningOnAzurePipelinesHosted should be deprecated. If I wanted to detect the Environment I was in, I should just be checking Environment Variables. And if I wanted to know if I was hosted, I could just hit Environment.Agent.IsHosted. The value that this API provides is in me not knowing to have to check for "TF_BUILD" in the environment variables.

So even if IsRunningOnAzurePipelinesHosted not deprecated, IsRunningOnAzurePipelines should always return true if the build is running on Azure Pipelines, and !IsHosted should be removed from the check. I mean, it's the obvious semantic meaning of the words "Is Running On Azure Pipelines". 

As it is, it's not at all obvious that in order to detect if I'm running on Azure Pipelines, I _actually_ need to put  (IsRunningOnAzurePipelines || IsRunningOnAzurePipelinesHosted) everywhere instead of just IsRunningOnAzurePipelines.

This is one of those things that Cake can't check locally, which means that trying to debug it is a real pain.

**Comment by pascalberger** [comment]
Different people have different use-cases. Current implementation works fine for people which use hosted and on-prem pools and want to do something different. I agree that current properties don't work very well if you run on different CI servers and want to know if you are running on Azure Pipelines, but don't care if agent is running hosted or on-prem.

The current naming of `IsRunningOnAzurePipelines` and `IsRunningOnAzurePipelinesHosted` also no longer matches naming used by Microsoft which is now `Self hosted agent` and `Microsoft hosted agent`. Therefore it probably would be to deprecate the current properties and introduce new ones:

* IsRunningOnAzurePipelines -> IsRunningOnAzurePipelinesSelfHostedAgent
* IsRunningOnAzurePipelinesHosted -> IsRunningOnAzurePipelinesMicrosoftHostedAgent

If we use `IsRunningOnAzurePipelines` to indicate if running on either of the above we would change the meaning of the `IsRunningOnAzurePipelines` property in Cake 1.x to something different and break behavior of existing builds. So maybe this is something we can do in steps: 

* In Cake 2.0 introduce new properties to make it clear what they return
* In Cake 3.0 take care of the use case for an easy check if running on Azure Pipelines

There are also other issue with build system integration which might need some work or bigger refactoring, but it is not decided yet when this will happen, but for sure not before Cake 3.0.

**Reply by mandalorianbob** [reply]
I think the important information that IsRunningOnAzurePipelines provides is the environment variable check. There definitely needs to be a property that just returns that information alone. IsRunningInAzurePipelinesEnvironment maybe.

Something I should mention is that the landscape of hosting providers is changing. So self-hosted and Microsoft hosted are getting even fuzzier, and "self-hosted" doesn't even mean on premises necessarily. 

In the meanwhile, documentation should probably be updated. "Gets a value indicating whether the current build is running on Azure Pipelines." is obviously insufficient. It should probably mention that it's only checking if it's running on a self-hosted Azure Pipelines build.

**Comment by pascalberger** [comment]
This will be tracked with https://github.com/cake-build/cake/issues/3654
